core/lf_token.c

```diff
@@ -0,0 +1,358 @@
+ *    this list of conditions and the following disclaimer.
+ *
+ * 2. Redistributions in binary form must reproduce the above copyright notice,
+ *    this list of conditions and the following disclaimer in the documentation
+ *    and/or other materials provided with the distribution.
+ *
+ * THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS" AND ANY
+ * EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE IMPLIED WARRANTIES OF
+ * MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL
+ * THE COPYRIGHT HOLDER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL,
+ * SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO,
+ * PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS
+ * INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT,
+ * STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF
+ * THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+ *
+ * @section DESCRIPTION
+ *
+ * Functions supporting token types.  See lf_token.h for docs.
+ */
+
+#include <stdbool.h>
+#include <assert.h>
+#include <string.h>  // Defines memcpy
+#include "lf_token.h"
+#include "hashset/hashset_itr.h"
+#include "util.h"
+#include "platform.h" // Defines lf_critical_section_enter() and exit.
+#include "port.h"     // Defines lf_port_base_t.
+
+lf_token_t* _lf_tokens_allocated_in_reactions = NULL;
+
+////////////////////////////////////////////////////////////////////
+//// Global variables not visible outside this file.
+
+/**
+ * Tokens always have the same size in memory so they are easily recycled.
+ * When a token is freed, it is inserted into this recycling bin.
+ */
+static hashset_t _lf_token_recycling_bin = NULL;
+
+/**
+ * To allow a system to recover from burst of activity, the token recycling
+ * bin has a limited size. When it becomes full, token are freed using free().
+ */
+#define _LF_TOKEN_RECYCLING_BIN_SIZE_LIMIT 512
+
+/**
+ * Set of token templates (trigger_t or port_base_t objects) that
+ * have been initialized. This is used to free their tokens at
+ * the end of program execution. 
+ */
+static hashset_t _lf_token_templates = NULL;
+
+////////////////////////////////////////////////////////////////////
+//// Functions that user's may call.
+
+lf_token_t* lf_new_token(void* port_or_action, void* val, size_t len) {
+    return _lf_new_token((token_type_t*)port_or_action, val, len);
+}
+
+lf_token_t* lf_writable_copy(lf_port_base_t* port) {
+    assert(port != NULL);
+
+    lf_token_t* token = port->tmplt.token;
+    if (token == NULL) return NULL;
+    LF_PRINT_DEBUG("lf_writable_copy: Requesting writable copy of token %p with reference count %zu.",
+            token, token->ref_count);
+    // FIXME: token_template_t needs to be augmented with a field
```

Comment:
Please document this in an issue and reference it in the code.

Reply:
Actually, this FIXME is fixed.  Removed the comment.